File: bot/config/chains.py
from dataclasses import dataclass
from typing import Optional, Union
from enum import Enum


class ChainType(Enum):
    """Blockchain type."""

    EVM = "evm"
    SOLANA = "solana"
    TRON = "tron"
    STARKNET = "starknet"


@dataclass
class ChainConfig:
    """Configuration for a blockchain network."""

    chain_id: Union[int, str]  # int for EVM, str for Solana
    name: str
    display_name: str
    chain_type: ChainType
    native_token: str
    native_decimals: int
    rpc_url_env: str  # Environment variable name for RPC URL
    explorer_url: str
    logo_emoji: str
    lifi_chain_id: Optional[int] = None  # Li.Fi uses numeric chain IDs
    # Chains with no EIP-1559 support must use legacy gasPrice transactions only.
    # (Our EVM send path is already legacy-gasPrice everywhere; this flag documents
    # the constraint and guards any future EIP-1559 migration.)
    legacy_gas_only: bool = False
    # Network-enforced minimum gas price in wei (e.g. Rootstock's 0.06 gwei floor).
    # Applied via apply_min_gas_price() in the tx-build path.
    min_gas_price_wei: int = 0


# Chain configurations
CHAINS: dict[str, ChainConfig] = {
    "ethereum": ChainConfig(
        chain_id=1,
        name="ethereum",
        display_name="ETH",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="ETHEREUM_RPC_URL",
        explorer_url="https://etherscan.io",
        logo_emoji="🔷",
        lifi_chain_id=1,
    ),
    "bsc": ChainConfig(
        chain_id=56,
        name="bsc",
        display_name="BSC",
        chain_type=ChainType.EVM,
        native_token="BNB",
        native_decimals=18,
        rpc_url_env="BSC_RPC_URL",
        explorer_url="https://bscscan.com",
        logo_emoji="🟡",
        lifi_chain_id=56,
    ),
    "polygon": ChainConfig(
        chain_id=137,
        name="polygon",
        display_name="POL",
        chain_type=ChainType.EVM,
        native_token="MATIC",
        native_decimals=18,
        rpc_url_env="POLYGON_RPC_URL",
        explorer_url="https://polygonscan.com",
        logo_emoji="🟣",
        lifi_chain_id=137,
    ),
    "arbitrum": ChainConfig(
        chain_id=42161,
        name="arbitrum",
        display_name="ARB",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="ARBITRUM_RPC_URL",
        explorer_url="https://arbiscan.io",
        logo_emoji="🔵",
        lifi_chain_id=42161,
    ),
    "optimism": ChainConfig(
        chain_id=10,
        name="optimism",
        display_name="OP",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="OPTIMISM_RPC_URL",
        explorer_url="https://optimistic.etherscan.io",
        logo_emoji="🔴",
        lifi_chain_id=10,
    ),
    "base": ChainConfig(
        chain_id=8453,
        name="base",
        display_name="Base",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="BASE_RPC_URL",
        explorer_url="https://basescan.org",
        logo_emoji="🔵",
        lifi_chain_id=8453,
    ),
    "base-sepolia": ChainConfig(
        chain_id=84532,
        name="base-sepolia",
        display_name="Base Sepolia",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="BASE_SEPOLIA_RPC_URL",
        explorer_url="https://sepolia.basescan.org",
        logo_emoji="🔵",
        lifi_chain_id=None,  # Li.Fi has no testnet support — native escrow / direct transfers only
    ),
    "avalanche": ChainConfig(
        chain_id=43114,
        name="avalanche",
        display_name="AVAX",
        chain_type=ChainType.EVM,
        native_token="AVAX",
        native_decimals=18,
        rpc_url_env="AVALANCHE_RPC_URL",
        explorer_url="https://snowtrace.io",
        logo_emoji="🔺",
        lifi_chain_id=43114,
    ),
    "fantom": ChainConfig(
        chain_id=250,
        name="fantom",
        display_name="FTM",
        chain_type=ChainType.EVM,
        native_token="FTM",
        native_decimals=18,
        rpc_url_env="FANTOM_RPC_URL",
        explorer_url="https://ftmscan.com",
        logo_emoji="👻",
        lifi_chain_id=250,
    ),
    "linea": ChainConfig(
        chain_id=59144,
        name="linea",
        display_name="Linea",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="LINEA_RPC_URL",
        explorer_url="https://lineascan.build",
        logo_emoji="🟦",
        lifi_chain_id=59144,
    ),
    "mantle": ChainConfig(
        chain_id=5000,
        name="mantle",
        display_name="Mantle",
        chain_type=ChainType.EVM,
        native_token="MNT",
        native_decimals=18,
        rpc_url_env="MANTLE_RPC_URL",
        explorer_url="https://mantlescan.xyz",
        logo_emoji="🟩",
        lifi_chain_id=5000,
    ),
    "gnosis": ChainConfig(
        chain_id=100,
        name="gnosis",
        display_name="GNO",
        chain_type=ChainType.EVM,
        native_token="xDAI",
        native_decimals=18,
        rpc_url_env="GNOSIS_RPC_URL",
        explorer_url="https://gnosisscan.io",
        logo_emoji="🦉",
        lifi_chain_id=100,
    ),
    "scroll": ChainConfig(
        chain_id=534352,
        name="scroll",
        display_name="Scroll",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="SCROLL_RPC_URL",
        explorer_url="https://scrollscan.com",
        logo_emoji="📜",
        lifi_chain_id=534352,
    ),
    "solana": ChainConfig(
        chain_id="solana",
        name="solana",
        display_name="SOL",
        chain_type=ChainType.SOLANA,
        native_token="SOL",
        native_decimals=9,
        rpc_url_env="SOLANA_RPC_URL",
        explorer_url="https://solscan.io",
        logo_emoji="🟢",
        lifi_chain_id=1151111081099710,  # Li.Fi Solana chain ID
    ),
    "tron": ChainConfig(
        chain_id="tron",
        name="tron",
        display_name="TRX",
        chain_type=ChainType.TRON,
        native_token="TRX",
        native_decimals=6,
        rpc_url_env="TRON_RPC_URL",
        explorer_url="https://tronscan.org",
        logo_emoji="💎",
        lifi_chain_id=1170,
    ),
    "starknet": ChainConfig(
        chain_id="SN_MAIN",
        name="starknet",
        display_name="Starknet",
        chain_type=ChainType.STARKNET,
        native_token="STRK",
        native_decimals=18,
        rpc_url_env="STARKNET_RPC_URL",
        explorer_url="https://voyager.online",
        logo_emoji="🐺",
        lifi_chain_id=None,  # Li.Fi does not support Starknet
    ),
    "tempo": ChainConfig(
        chain_id=4217,
        name="tempo",
        display_name="Tempo",
        chain_type=ChainType.EVM,
        native_token="USD",
        native_decimals=6,
        rpc_url_env="TEMPO_RPC_URL",
        explorer_url="https://explore.tempo.xyz",
        logo_emoji="⚡",
        lifi_chain_id=4217,
    ),
    "plasma": ChainConfig(
        chain_id=9745,
        name="plasma",
        display_name="Plasma",
        chain_type=ChainType.EVM,
        native_token="XPL",
        native_decimals=18,
        rpc_url_env="PLASMA_RPC_URL",
        explorer_url="https://plasmascan.to",
        logo_emoji="🟪",
        lifi_chain_id=9745,
    ),
    "goat": ChainConfig(
        chain_id=2345,
        name="goat",
        display_name="GOAT",
        chain_type=ChainType.EVM,
        native_token="BTC",
        native_decimals=18,  # GOAT's native BTC uses 18 decimals (ETH-style), NOT 8
        rpc_url_env="GOAT_RPC_URL",
        explorer_url="https://explorer.goat.network",
        logo_emoji="🐐",
        lifi_chain_id=None,  # Li.Fi does not support GOAT — routes via GOATSwap only
    ),
    "rootstock": ChainConfig(
        chain_id=30,
        name="rootstock",
        display_name="Rootstock",
        chain_type=ChainType.EVM,
        native_token="RBTC",
        native_decimals=18,
        rpc_url_env="ROOTSTOCK_RPC_URL",
        explorer_url="https://rootstock.blockscout.com",
        logo_emoji="🟧",
        # Routing: Li.Fi ONLY (chain 30, verified live). Rootstock must NOT be added
        # to the 1inch/0x/Kyber/OKX/Across/CCTP chain dicts — absence = excluded.
        lifi_chain_id=30,
        # Rootstock has no EIP-1559 (eth_feeHistory unsupported) and uses
        # EIP-1191 chain-salted checksums — compare addresses lowercased,
        # never validate by EIP-55 checksum.
        legacy_gas_only=True,
        min_gas_price_wei=60_000_000,  # 0.06 gwei network minimum
    ),
    "citrea": ChainConfig(
        chain_id=4114,
        name="citrea",
        display_name="Citrea",
        chain_type=ChainType.EVM,
        native_token="cBTC",
        native_decimals=18,  # native cBTC uses 18 decimals (ETH-style), NOT 8
        rpc_url_env="CITREA_RPC_URL",
        explorer_url="https://explorer.mainnet.citrea.xyz",
        logo_emoji="🍊",
        # Routing: JuiceSwap (direct UniV3 fork) ONLY — Citrea (chain id 4114)
        # is absent from EVERY aggregator (LiFi/1inch/0x/Kyber/OKX/CoW/Socket).
        lifi_chain_id=None,
        # EIP-1559 OK (Type-2 zkEVM, Pectra), but the L1 (Bitcoin DA) fee
        # surcharge is NOT included in eth_estimateGas — execution adds 15%
        # headroom (see univ3_fork_api.CITREA_VENUE.gas_headroom_pct).
    ),
    # === New Li.Fi-supported chains (2025-2026) ===
    "sonic": ChainConfig(
        chain_id=146,
        name="sonic",
        display_name="Sonic",
        chain_type=ChainType.EVM,
        native_token="S",
        native_decimals=18,
        rpc_url_env="SONIC_RPC_URL",
        explorer_url="https://sonicscan.org",
        logo_emoji="🔵",
        lifi_chain_id=146,
    ),
    "opbnb": ChainConfig(
        chain_id=204,
        name="opbnb",
        display_name="opBNB",
        chain_type=ChainType.EVM,
        native_token="BNB",
        native_decimals=18,
        rpc_url_env="OPBNB_RPC_URL",
        explorer_url="https://opbnb.bscscan.com",
        logo_emoji="🟡",
        lifi_chain_id=204,
    ),
    "fraxtal": ChainConfig(
        chain_id=252,
        name="fraxtal",
        display_name="Fraxtal",
        chain_type=ChainType.EVM,
        native_token="FRAX",
        native_decimals=18,
        rpc_url_env="FRAXTAL_RPC_URL",
        explorer_url="https://fraxscan.com",
        logo_emoji="⚫",
        lifi_chain_id=252,
    ),
    "zksync": ChainConfig(
        chain_id=324,
        name="zksync",
        display_name="zkSync Era",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="ZKSYNC_RPC_URL",
        explorer_url="https://explorer.zksync.io",
        logo_emoji="🔷",
        lifi_chain_id=324,
    ),
    "worldchain": ChainConfig(
        chain_id=480,
        name="worldchain",
        display_name="World Chain",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="WORLDCHAIN_RPC_URL",
        explorer_url="https://worldscan.org",
        logo_emoji="🌍",
        lifi_chain_id=480,
    ),
    "flow": ChainConfig(
        chain_id=747,
        name="flow",
        display_name="Flow",
        chain_type=ChainType.EVM,
        native_token="FLOW",
        native_decimals=18,
        rpc_url_env="FLOW_RPC_URL",
        explorer_url="https://evm.flowscan.io",
        logo_emoji="🟢",
        lifi_chain_id=747,
    ),
    "hyperevm": ChainConfig(
        chain_id=999,
        name="hyperevm",
        display_name="HyperEVM",
        chain_type=ChainType.EVM,
        native_token="HYPE",
        native_decimals=18,
        rpc_url_env="HYPEREVM_RPC_URL",
        explorer_url="https://explorer.hyperliquid.xyz",
        logo_emoji="⚡",
        lifi_chain_id=999,
    ),
    "lisk": ChainConfig(
        chain_id=1135,
        name="lisk",
        display_name="Lisk",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="LISK_RPC_URL",
        explorer_url="https://blockscout.lisk.com",
        logo_emoji="🔵",
        lifi_chain_id=1135,
    ),
    "sei": ChainConfig(
        chain_id=1329,
        name="sei",
        display_name="Sei",
        chain_type=ChainType.EVM,
        native_token="SEI",
        native_decimals=18,
        rpc_url_env="SEI_RPC_URL",
        explorer_url="https://seitrace.com",
        logo_emoji="🔴",
        lifi_chain_id=1329,
    ),
    "soneium": ChainConfig(
        chain_id=1868,
        name="soneium",
        display_name="Soneium",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="SONEIUM_RPC_URL",
        explorer_url="https://soneium.blockscout.com",
        logo_emoji="🟦",
        lifi_chain_id=1868,
    ),
    "swellchain": ChainConfig(
        chain_id=1923,
        name="swellchain",
        display_name="Swellchain",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="SWELLCHAIN_RPC_URL",
        explorer_url="https://explorer.swellnetwork.io",
        logo_emoji="🌊",
        lifi_chain_id=1923,
    ),
    "abstract": ChainConfig(
        chain_id=2741,
        name="abstract",
        display_name="Abstract",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="ABSTRACT_RPC_URL",
        explorer_url="https://abscan.org",
        logo_emoji="🟪",
        lifi_chain_id=2741,
    ),
    "kaia": ChainConfig(
        chain_id=8217,
        name="kaia",
        display_name="Kaia",
        chain_type=ChainType.EVM,
        native_token="KAIA",
        native_decimals=18,
        rpc_url_env="KAIA_RPC_URL",
        explorer_url="https://kaiascan.io",
        logo_emoji="🟠",
        lifi_chain_id=8217,
    ),
    "apechain": ChainConfig(
        chain_id=33139,
        name="apechain",
        display_name="Apechain",
        chain_type=ChainType.EVM,
        native_token="APE",
        native_decimals=18,
        rpc_url_env="APECHAIN_RPC_URL",
        explorer_url="https://apescan.io",
        logo_emoji="🦍",
        lifi_chain_id=33139,
    ),
    "mode": ChainConfig(
        chain_id=34443,
        name="mode",
        display_name="Mode",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="MODE_RPC_URL",
        explorer_url="https://modescan.io",
        logo_emoji="🟡",
        lifi_chain_id=34443,
    ),
    "hemi": ChainConfig(
        chain_id=43111,
        name="hemi",
        display_name="Hemi",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="HEMI_RPC_URL",
        explorer_url="https://explorer.hemi.xyz",
        logo_emoji="🟠",
        lifi_chain_id=43111,
    ),
    "bob": ChainConfig(
        chain_id=60808,
        name="bob",
        display_name="BOB",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="BOB_RPC_URL",
        explorer_url="https://explorer.gobob.xyz",
        logo_emoji="🟡",
        lifi_chain_id=60808,
    ),
    "berachain": ChainConfig(
        chain_id=80094,
        name="berachain",
        display_name="Berachain",
        chain_type=ChainType.EVM,
        native_token="BERA",
        native_decimals=18,
        rpc_url_env="BERACHAIN_RPC_URL",
        explorer_url="https://berascan.com",
        logo_emoji="🐻",
        lifi_chain_id=80094,
    ),
    "taiko": ChainConfig(
        chain_id=167000,
        name="taiko",
        display_name="Taiko",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="TAIKO_RPC_URL",
        explorer_url="https://taikoscan.io",
        logo_emoji="🥁",
        lifi_chain_id=167000,
    ),
    "unichain": ChainConfig(
        chain_id=130,
        name="unichain",
        display_name="Unichain",
        chain_type=ChainType.EVM,
        native_token="ETH",
        native_decimals=18,
        rpc_url_env="UNICHAIN_RPC_URL",
        explorer_url="https://uniscan.xyz",
        logo_emoji="🦄",
        lifi_chain_id=130,
    ),
    "flare": ChainConfig(
        chain_id=14,
        name="flare",
        display_name="Flare",
        chain_type=ChainType.EVM,
        native_token="FLR",
        native_decimals=18,
        rpc_url_env="FLARE_RPC_URL",
        explorer_url="https://flarescan.com",
        logo_emoji="🔥",
        lifi_chain_id=14,
    ),
}


# Tempo testnets — intentionally NOT in CHAINS (not user-selectable in the bot).
# For tooling / integration tests only. Verified against tempoxyz docs + ChainList.
# The enshrined DEX (0xDEc0…), TIP-20 factory (0x20Fc…), and the stablecoin
# precompiles (0x20C0…0000-0003) are system contracts identical to mainnet, so the
# same client code (bot/services/tempo_*.py) works against these RPCs unchanged.
TEMPO_TESTNETS: dict[str, dict] = {
    # Current primary testnet (chain id confirmed via tempoxyz/tempo + ChainList).
    "moderato": {
        "chain_id": 42431,
        "rpc_url": "https://rpc.moderato.tempo.xyz",
        "explorer_url": "https://explore.tempo.xyz",
    },
    # Earlier testnet; RPC reachable via thirdweb's proxy (no canonical
    # *.tempo.xyz host published, so we don't guess one).
    "andantino": {
        "chain_id": 42429,
        "rpc_url": "https://42429.rpc.thirdweb.com",
        "explorer_url": "https://explore.tempo.xyz",
    },
}


def apply_min_gas_price(chain_name: str, gas_price: int) -> int:
    """Enforce a chain's network-minimum gas price (wei) on a fetched gas price.

    Rootstock rejects txs below 60M wei (0.06 gwei); other chains pass through.
    """
    chain = CHAINS.get(chain_name.lower())
    if chain and chain.min_gas_price_wei:
        return max(int(gas_price), chain.min_gas_price_wei)
    return int(gas_price)


def get_chain_by_id(chain_id: Union[int, str]) -> Optional[ChainConfig]:
    """Get chain configuration by chain ID."""
    for chain in CHAINS.values():
        if chain.chain_id == chain_id:
            return chain
    return None


def get_chain_by_name(name: str) -> Optional[ChainConfig]:
    """Get chain configuration by name."""
    return CHAINS.get(name.lower())


def get_evm_chains() -> list[ChainConfig]:
    """Get all EVM chain configurations."""
    return [c for c in CHAINS.values() if c.chain_type == ChainType.EVM]


def get_solana_chain() -> ChainConfig:
    """Get Solana chain configuration."""
    return CHAINS["solana"]


def get_tron_chain() -> ChainConfig:
    """Get TRON chain configuration."""
    return CHAINS["tron"]


def get_starknet_chain() -> ChainConfig:
    """Get Starknet chain configuration."""
    return CHAINS["starknet"]

File: webapp/src/lib/chains.ts
// Display-only chain metadata for the webapp (names, icon fallbacks, explorer links).
// Keep in sync with api-ts/src/config/chains.ts and bot/config/chains.py — this module
// must stay presentation-only (no RPC URLs, no signing logic).

export interface ChainDisplay {
  /** Human-readable chain name */
  name: string
  /** Emoji fallback icon (webapp has no chain image assets) */
  icon: string
  /** Explorer base URL, no trailing slash. Absent when we don't know the chain's explorer. */
  explorerUrl?: string
}

export const CHAIN_DISPLAY: Record<string, ChainDisplay> = {
  ethereum: { name: 'Ethereum', icon: '⟠', explorerUrl: 'https://etherscan.io' },
  optimism: { name: 'Optimism', icon: '\u{1F534}', explorerUrl: 'https://optimistic.etherscan.io' },
  bsc: { name: 'BNB Chain', icon: '\u{1F7E1}', explorerUrl: 'https://bscscan.com' },
  polygon: { name: 'Polygon', icon: '\u{1F7E3}', explorerUrl: 'https://polygonscan.com' },
  base: { name: 'Base', icon: '\u{1F535}', explorerUrl: 'https://basescan.org' },
  'base-sepolia': { name: 'Base Sepolia', icon: '\u{1F9EA}', explorerUrl: 'https://sepolia.basescan.org' },
  arbitrum: { name: 'Arbitrum', icon: '\u{1F537}', explorerUrl: 'https://arbiscan.io' },
  solana: { name: 'Solana', icon: '\u{1F30A}', explorerUrl: 'https://solscan.io' },
  tempo: { name: 'Tempo', icon: '⏱️', explorerUrl: 'https://explore.tempo.xyz' },
  robinhood: { name: 'Robinhood', icon: '🪶', explorerUrl: 'https://robinhoodchain.blockscout.com' },
  plasma: { name: 'Plasma', icon: '⚡', explorerUrl: 'https://plasmascan.to' },
  starknet: { name: 'Starknet', icon: '\u{1F680}', explorerUrl: 'https://voyager.online' },
  goat: { name: 'GOAT Network', icon: '\u{1F410}', explorerUrl: 'https://explorer.goat.network' },
  avalanche: { name: 'Avalanche', icon: '\u{1F53A}', explorerUrl: 'https://snowtrace.io' },
  sepolia: { name: 'Sepolia', icon: '⟠', explorerUrl: 'https://sepolia.etherscan.io' },
  fantom: { name: 'Fantom', icon: '\u{1F47B}', explorerUrl: 'https://ftmscan.com' },
  linea: { name: 'Linea', icon: '\u{1F9F5}', explorerUrl: 'https://lineascan.build' },
  mantle: { name: 'Mantle', icon: '\u{1F3D4}️', explorerUrl: 'https://mantlescan.xyz' },
  gnosis: { name: 'Gnosis', icon: '\u{1F989}', explorerUrl: 'https://gnosisscan.io' },
  scroll: { name: 'Scroll', icon: '\u{1F4DC}', explorerUrl: 'https://scrollscan.com' },
  tron: { name: 'Tron', icon: '\u{1F53B}', explorerUrl: 'https://tronscan.org' },
  rootstock: { name: 'Rootstock', icon: '\u{1FAA8}', explorerUrl: 'https://rootstock.blockscout.com' },
  citrea: { name: 'Citrea', icon: '\u{1F34A}', explorerUrl: 'https://explorer.mainnet.citrea.xyz' },
  sonic: { name: 'Sonic', icon: '\u{1F4A8}', explorerUrl: 'https://sonicscan.org' },
  opbnb: { name: 'opBNB', icon: '\u{1F7E8}', explorerUrl: 'https://opbnb.bscscan.com' },
  fraxtal: { name: 'Fraxtal', icon: '\u{1F9CA}', explorerUrl: 'https://fraxscan.com' },
  zksync: { name: 'zkSync Era', icon: '\u{1F512}', explorerUrl: 'https://explorer.zksync.io' },
  worldchain: { name: 'World Chain', icon: '\u{1F30D}', explorerUrl: 'https://worldscan.org' },
  flow: { name: 'Flow', icon: '\u{1F300}', explorerUrl: 'https://evm.flowscan.io' },
  hyperevm: { name: 'HyperEVM', icon: '\u{1F985}', explorerUrl: 'https://explorer.hyperliquid.xyz' },
  lisk: { name: 'Lisk', icon: '\u{1F331}', explorerUrl: 'https://blockscout.lisk.com' },
  sei: { name: 'Sei', icon: '\u{1F40B}', explorerUrl: 'https://seitrace.com' },
  soneium: { name: 'Soneium', icon: '\u{1F3B5}', explorerUrl: 'https://soneium.blockscout.com' },
  swellchain: { name: 'Swellchain', icon: '\u{1FAE7}', explorerUrl: 'https://explorer.swellnetwork.io' },
  abstract: { name: 'Abstract', icon: '\u{1F3A8}', explorerUrl: 'https://abscan.org' },
  kaia: { name: 'Kaia', icon: '\u{1F33F}', explorerUrl: 'https://kaiascan.io' },
  apechain: { name: 'Apechain', icon: '\u{1F412}', explorerUrl: 'https://apescan.io' },
  mode: { name: 'Mode', icon: '\u{1F39B}️', explorerUrl: 'https://modescan.io' },
  hemi: { name: 'Hemi', icon: '\u{1F317}', explorerUrl: 'https://explorer.hemi.xyz' },
  bob: { name: 'BOB', icon: '\u{1F171}️', explorerUrl: 'https://explorer.gobob.xyz' },
  berachain: { name: 'Berachain', icon: '\u{1F43B}', explorerUrl: 'https://berascan.com' },
  taiko: { name: 'Taiko', icon: '\u{1F941}', explorerUrl: 'https://taikoscan.io' },
  unichain: { name: 'Unichain', icon: '\u{1F984}', explorerUrl: 'https://uniscan.xyz' },
  flare: { name: 'Flare', icon: '\u{1F525}', explorerUrl: 'https://flarescan.com' },
  aurora: { name: 'Aurora', icon: '\u{1F30C}', explorerUrl: 'https://explorer.aurora.dev' },
  blast: { name: 'Blast', icon: '\u{1F4A5}', explorerUrl: 'https://blastscan.io' },
  ink: { name: 'Ink', icon: '\u{1F58B}️', explorerUrl: 'https://explorer.inkonchain.com' },
}

/** Resolve display metadata for a chain key (case-insensitive); falls back to an unbranded entry
 *  with no explorer URL — callers must NOT assume `explorerUrl` is set for unknown chains. */
export function getChainDisplay(chain: string): ChainDisplay {
  const key = chain.toLowerCase().trim()
  return (
    CHAIN_DISPLAY[key] ?? {
      name: chain,
      icon: '\u{1F517}',
    }
  )
}

/** Build an explorer transaction link for a chain key, or `null` if the chain's explorer
 *  is unknown. Callers must handle the `null` case (e.g. render no link) rather than
 *  falling back to a guessed explorer — a wrong link is worse than no link. */
export function getExplorerTxUrl(chain: string, txHash: string): string | null {
  const { explorerUrl } = getChainDisplay(chain)
  return explorerUrl ? `${explorerUrl}/tx/${txHash}` : null
}

/** Numeric EVM chain id (as a string) -> CHAIN_DISPLAY key. Only covers chains with a
 *  numeric chain id (EVM chains + the sepolia testnet); non-EVM chains (solana, tron,
 *  starknet) have no numeric id and are looked up by name directly. */
export const CHAIN_ID_TO_KEY: Record<string, string> = {
  '1': 'ethereum',
  '10': 'optimism',
  '14': 'flare',
  '30': 'rootstock',
  '56': 'bsc',
  '100': 'gnosis',
  '130': 'unichain',
  '137': 'polygon',
  '146': 'sonic',
  '204': 'opbnb',
  '250': 'fantom',
  '252': 'fraxtal',
  '324': 'zksync',
  '480': 'worldchain',
  '747': 'flow',
  '999': 'hyperevm',
  '1135': 'lisk',
  '1329': 'sei',
  '1868': 'soneium',
  '1923': 'swellchain',
  '2345': 'goat',
  '2741': 'abstract',
  '4114': 'citrea',
  '4217': 'tempo',
  '4663': 'robinhood',
  '5000': 'mantle',
  '8217': 'kaia',
  '8453': 'base',
  '9745': 'plasma',
  '11155111': 'sepolia',
  '33139': 'apechain',
  '34443': 'mode',
  '42161': 'arbitrum',
  '43111': 'hemi',
  '43114': 'avalanche',
  '57073': 'ink',
  '59144': 'linea',
  '60808': 'bob',
  '80094': 'berachain',
  '81457': 'blast',
  '84532': 'base-sepolia',
  '167000': 'taiko',
  '534352': 'scroll',
  '1313161554': 'aurora',
}

/** Resolve a CHAIN_DISPLAY key from a numeric EVM chain id (string or number). Returns
 *  `undefined` for unknown or non-EVM ids — do NOT default to 'ethereum'. */
export function getChainKeyByChainId(chainId: string | number): string | undefined {
  return CHAIN_ID_TO_KEY[String(chainId)]
}

/** Build an explorer transaction link from a numeric EVM chain id, or `null` if the
 *  chain id is unrecognized or its explorer is unknown. */
export function getExplorerTxUrlByChainId(chainId: string | number, txHash: string): string | null {
  const key = getChainKeyByChainId(chainId)
  return key ? getExplorerTxUrl(key, txHash) : null
}
